src/psyclone/psyGen.py

```diff
@@ -2659,7 +2669,28 @@ def _get_private_list(self):
             # We have at least two accesses. If the first one is a write,
             # assume the variable should be private:
             if accesses[0].access_type == AccessType.WRITE:
-                result.add(var_name.lower())
+                # Check if the write access is inside the parallel loop. If
+                # the write is outside of a loop, it is an assignment to
+                # a shared variable. Example where jpk is likely used
+                # outside of the parallel section later, so it must be
+                # declared as shared in order to have its value in other loops:
+                # !$omp parallel
+                # jpk = 100
+                # !omp do
+                # do ji = 1, jpk
+
+                # Go up the tree till we either find the InvokeSchedule,
+                # which is at the top, or a Loop statement (or no parent,
+                # which means we have reached the end of a called kernel).
+                parent = accesses[0].node
+                while parent and \
```

Comment:
Looking at this afresh, I think you could use the `Node.ancestor()` method to achieve the same thing.

Reply:
No, ancestor() does not return self (if applicable). I tried to support returning self in ancestor, but then the read_ancestor_test failed, so I assume it is important that ancestor does not return self. And I indeed need to be able to find self.

Reply:
Would you mind creating a new issue for this and mark this bit of code with a TODO?

Reply:
Done: #597.